### ISSUE TITLE
storage: orig range is not splittable by new range

### ISSUE THREAD

**Issue by danhhz** [cockroachdb/cockroach#16008]
I just got this while testing some backup/restore stuff on lapis. I haven't had time to dig into it, but I'll save the logs.

```
F170517 20:22:31.425812 122 storage/store.go:1846  [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}] [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}]: failed to update Store after split: orig range is not splittable by new range: r1778:/Table/8{8/1/-2861451
5605402811-9/1} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4], r1792:/Table/8{8/1-9/1} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4]
```

https://sentry.io/cockroach-labs/cockroachdb/issues/275328817/

```
*errors.errorString: store.go:1846 %s: failed to update Store after split: %s
  File "github.com/cockroachdb/cockroach/pkg/storage/store.go", line 1846, in splitPostApply
  File "github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go", line 574, in handleReplicatedEvalResult
  File "github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go", line 794, in handleEvalResultRaftMuLocked
  File "github.com/cockroachdb/cockroach/pkg/storage/replica.go", line 3702, in processRaftCommand
  File "github.com/cockroachdb/cockroach/pkg/storage/replica.go", line 2872, in handleRaftReadyRaftMuLocked
...
(5 additional frame(s) were not displayed)

store.go:1846 %s: failed to update Store after split: %s
```

**Comment by danhhz** [comment]
Logs are on each lapis machine at ~/16008.tgz

@petermattis anything else I should save here? I'd like to get lapis back

**Comment by petermattis** [comment]
@danhhz Do you know the SHA this happened on? Wonder if it was related to https://github.com/cockroachdb/cockroach/issues/15979 and perhaps fixed by https://github.com/cockroachdb/cockroach/issues/15995. Seems like a long shot, but I don't recall seeing this error before.

Cc @bdarnell

**Comment by danhhz** [comment]
Looks like sentry actually has the sha. How cool! db8ec7437ecfa2d61f3b5cc726bbb5df5d5d0e16 (which does contain the fix you mention #15995)

**Comment by a-robinson** [comment]
Being related to #15935 seems fairly likely

**Comment by danhhz** [comment]
This just happened again on lapis at sha d9c7ad9533a8c80051e7b75478f787a7a2985dd3

https://sentry.io/cockroach-labs/cockroachdb/issues/275328818/events/

**Comment by bdarnell** [comment]
Here are all the logs for `r1778` (all on node 1. Nodes 2 and 3 had the crash too, but nothing else):

```
602:I170517 20:22:30.875742 733370 storage/replica_command.go:2633  [n1,s1,r1775/1:/Table/{87-90/1}] initiating a split of this range at key /Table/88/1/-28614515605402811 [r1778]
606:I170517 20:22:30.977592 733512 storage/replica_command.go:2633  [n1,s1,r1778/1:/Table/{88/1/-286…-90/1}] initiating a split of this range at key /Table/89/1 [r1780]
608:I170517 20:22:31.025806 733605 storage/replica_command.go:2633  [n1,s1,r1778/1:/Table/{88/1/-286…-90/1}] initiating a split of this range at key /Table/88/1 [r1791]
609:I170517 20:22:31.031888 108 storage/replica_proposal.go:385  [n1,s1,r1778/1:/Table/{88/1/-286…-90/1}] new range lease repl=(n1,s1):1 start=1495051840.385341565,0 epo=23 pro=1495051840.385343365,0 following repl=(n0,s0):? start=0.000000000,0 exp=0.000000000,0
610:I170517 20:22:31.043308 733605 storage/replica_command.go:2633  [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}] initiating a split of this range at key /Table/88/1 [r1792]
611:I170517 20:22:31.089849 122 storage/replica_proposal.go:385  [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}] new range lease repl=(n1,s1):1 start=1495051840.385341565,0 epo=23 pro=1495051840.385343365,0 following repl=(n0,s0):? start=0.000000000,0 exp=0.000000000,0
612:E170517 20:22:31.425770 122 util/log/crash_reporting.go:153  [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}] Reported as error eb9f260e31fb4d37b90981aa04e9de33
613:F170517 20:22:31.425812 122 storage/store.go:1846  [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}] [n1,s1,r1778/1:/Table/8{8/1/-286…-9/1}]: failed to update Store after split: orig range is not splittable by new range: r1778:/Table/8{8/1/-28614515605402811-9/1} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4], r1792:/Table/8{8/1-9/1} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4]
```

It looks like there were several tables/databases created at once (with IDs 87-90), and the split queue is creating ranges for those tables concurrently with either manual splits into table 88 or writing enough data for there to be a size-based split. We see three different goroutines calling `adminSplitWithDescriptor`, which is surprising - other than manual splits (of which there is at most one in this log), I think all the splits should be happening on the split queue goroutine. 

In the third log line

    608:I170517 20:22:31.025806 733605 storage/replica_command.go:2633  [n1,s1,r1778/1:/Table/{88/1/-286…-90/1}] initiating a split of this range at key /Table/88/1 [r1791]

we see that the range is being asked to split at a key that is out of its bounds. But before we log the "initiating a split" message, we have already checked that the [key is contained in this range](https://github.com/cockroachdb/cockroach/blob/b59d25ec7ca2eda48ab17028045eec7b2ae258f2/pkg/storage/replica_command.go#L2586) 

This could be a preexisting race that we never saw before because in normal circumstances, most splits happen on the split queue and manual splits are generally done serially by some external process. The race requires both automatic and manual splits to occur simultaneously. My initial thought is that it's unlikely to be related to #15935 because `adminSplitWithDescriptor` has never held any locks. It's possible that changes to the locking during execution of `EndTransaction` have changed things, though. 

There's one thing here I don't understand: After we see in the logs that the split key is out of bounds, the transaction (which does a conditional put of the range descriptor) succeeds (or rather it goes far enough down the commit path to hit this assertion). There's a discrepancy between the in-memory range descriptor and the one on disk, but that should only be possible in one direction, with the in-memory one lagging the on-disk one. This looks like the opposite.

**Comment by bdarnell** [comment]
@danhhz Can you turn on `--vmodule=log=1` on lapis in case this happens again? (that's `storage/log.go`)

**Comment by mjibson** [comment]
I've turned on that logging option on lapis.

**Comment by mjibson** [comment]
This has happened again with vmodule enabled. Logs are on each machine at `~/16008-2.tgz`. Node 1 didn't crash (and is still up), but the others did. Node 1 is also in a pretty weird state where the web UI assets load but the API won't respond with details. But that may be expected in a non-quorum situation.

**Comment by bdarnell** [comment]
This time we're interested in range 83, with all the relevant log lines found on node 4:

```
1329:I170523 19:14:19.183716 12317060 storage/replica_command.go:2633  [n4,s4,r76/1:/Table/68/1/{-34476…-112563…}] initiating a split of this range at key /Table/68/1/-1140927181189170488 [r83]
1337:I170523 19:14:19.450170 59 storage/replica_command.go:3538  [replicate,n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] change replicas (REMOVE_REPLICA (n1,s1):3): read existing descriptor r83:/Table/68/1/{-1140927181189170488-11256385855049725} [(n4,s4):1, (n2,s2):2, (n1,s1):3, (n3,s3):4, next=5]
1338:I170523 19:14:19.479797 59 storage/log.go:43  [replicate,n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] Range Event: "remove", range: 83, info: updated_desc:<range_id:83 start_key:"\314\211\200\360*\234\306\n\241\272\310" end_key:"\314\211\374'\375\237\220XO\375" replicas:<node_id:4 store_id:4 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:4 > next_replica_id:5 > removed_replica:<node_id:1 store_id:1 replica_id:3 > 
1341:I170523 19:14:19.627062 12318778 storage/replica_command.go:2633  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] initiating a split of this range at key /Table/68/1 [r85]
1342:I170523 19:14:19.711652 12318765 storage/replica.go:2617  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] proposing REMOVE_REPLICA (n1,s1):3: [(n4,s4):1 (n2,s2):2 (n3,s3):4]
1344:I170523 19:14:19.818726 12318778 storage/replica_command.go:2633  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] initiating a split of this range at key /Table/68/1 [r86]
1345:I170523 19:14:19.827596 12318778 storage/log.go:43  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] Range Event: "split", range: 83, info: updated_desc:<range_id:83 start_key:"\314\211\200\360*\234\306\n\241\272\310" end_key:"\314\211" replicas:<node_id:4 store_id:4 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:4 > next_replica_id:5 > new_desc:<range_id:86 start_key:"\314\211" end_key:"\314\211\374'\375\237\220XO\375" replicas:<node_id:4 store_id:4 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:3 > next_replica_id:4 > 
1346:I170523 19:14:19.888934 152 storage/replica_proposal.go:385  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] new range lease repl=(n4,s4):1 start=1495566859.101644423,0 epo=3 pro=1495566859.101659023,0 following repl=(n0,s0):? start=0.000000000,0 exp=0.000000000,0
1349:E170523 19:14:20.209622 152 util/log/crash_reporting.go:171  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] Reported as error 5eb72151fb09465a8cabfda829d23d3a
1350:F170523 19:14:20.209642 152 storage/store.go:1854  [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}] [n4,s4,r83/1:/Table/68/1/{-11409…-112563…}]: failed to update Store after split: orig range is not splittable by new range: r83:/Table/68/1/{-1140927181189170488-11256385855049725} [(n4,s4):1, (n2,s2):2, (n3,s3):4, next=5], r86:/Table/68/1{-/11256385855049725} [(n4,s4):1, (n2,s2):2, (n3,s3):3, next=4]
```

This time two splits were initiated (at the same key). This key was out of bounds of r83 from the beginning; it was also out of bounds of r76 (from which r83 was created). This split point belongs in r73, the parent of r76 (`1296:I170523 19:14:18.705997 3796425 storage/replica_command.go:2633  [n2,s2,r73/2:/Table/6{7-8/1/2307…}] initiating a split of this range at key /Table/68/1/-3447647950472116397 [r76]`). 

Earlier I had hypothesized that there was a discrepancy between the on-disk and in-memory range descriptors that allowed the split transaction to proceed with this out-of-bounds key. That does not seem to fit with this new log, since the proposed split key was never valid in any version of r83. I don't understand how the split command could have been routed to this range.

**Comment by danhhz** [comment]
Any progress on this? It happens on every RESTORE of non-trivial size and so is blocking production testing

**Comment by bdarnell** [comment]
I'm looking into this today. The key validation at the start of AdminSplit is broken in several ways, most notably that it [transforms the key](https://github.com/cockroachdb/cockroach/blob/0677ecde3fc2e4b724d9b8d79b64340d01ca7921/pkg/storage/replica_command.go#L2624) with `EnsureSafeSplitKey` *after* it has validated it with `r.ContainsKey`. My working theory is that through some difference in the way split keys are chosen in restore compared to other splits, `EnsureSafeSplitKey` is removing too much from the end of the key and resulting in a value that is out of range.

**Comment by bdarnell** [comment]
#16291 improves the validation in AdminSplit. That'll give us a cleaner error message if the key encoding is going wrong somehow. I don't see how that could be happening since the restore code is correctly calling MakeRowSentinelKey, but I could be missing something. 

Another detail is that the splits generated by restore look like `/Table/68/1`: they're pointing to the beginning of the table's primary key. The splits generated by the regular split queue look like `/Table/68`, without any index id. I'm not sure where these keys are coming from.

**Comment by bdarnell** [comment]
@danhhz Have you tried a restore with #16291 yet? That may give us a better error about what's going on, or rule out this kind of key transformation as a problem.

The reason restore splits at points like `/Table/68/1` is that backups run at the *index* level rather than the *table* level. Backups don't scan the (empty) span between `/Table/68` and `/Table/68/1`, so restore tries to split at `/Table/68/1` while the split queue tries to split at `/Table/68`. This is not ideal because it creates an extra empty range, although it shouldn't break anything. It's possible that some of the assumptions in MakeRowSentinelKey and EnsureSafeSplitKey are broken by this unusual manual key generation, though. We should change EnsureSafeSplitKey into a read-only "check" function and make it the caller's responsibility to provide a valid split key, instead of taking what the caller provided and transforming it.

**Comment by danhhz** [comment]
There's a teamcity job that auto-runs a backup and restore every morning (https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Internal_Production_BackupToAzure). Looks like today's failed, though I haven't looked at the logs yet to see if it was this same failure

I agree that the extra empty range is unfortunate, we should look into that. I looked pretty closely at how we were doing the splits when this all first got written and didn't see any assumptions it was breaking. Maybe you see something I don't though?

This all works very well in 1.0 and the badness started without any relevant changes in backup/restore. Do you think we should track down when it broke?

**Comment by danhhz** [comment]
I just filed https://github.com/cockroachdb/cockroach/issues/16317 for the extra empty range

**Comment by bdarnell** [comment]
Today's run failed with the new message:

```
Error: pq: presplitting 21 ranges: storage/replica_command.go:2649: requested split key /Table/58/1/0 changed to /Table/58/1 by EnsureSafeSplitKey, out of bounds of [n1,s1,r19/1:/Table/58/1/{-48827���-115476���}]
```

This is interesting. `MakeRowSentinelKey(k)` is used as if it returns `k + epsilon`, but it actually returns `concat(k, 0)`, and zero is not the smallest possible integer. That's why the request is being sent to the wrong range. `MakeRowSentinelKey` is incorrect when applied to keys other than "row keys" like `/Table/68` and `/Table/68/1`. (This bug has always been around. We didn't see it pre-RESTORE because we never had so much parallelism around splits. I don't know why we didn't see this in 1.0 RESTORE testing. Maybe we've just improved parallelism further with changes like #15935?)

As a quick fix, I think we can change the client-side use of `AdminSplit` to use the desired split point as-is for `AdminSplitRequest.Span.Key` and the version transformed by `MakeRowSentinelKey` for `AdminSplitRequest.SplitKey`. Longer-term we need to refactor the way `MakeRowSentinelKey` is used, because many of these uses don't really make sense.

**Comment by danhhz** [comment]
#15935 is exactly the same time frame that we started seeing this, which lends some support to that theory.

@benesch Do you think you could pick off the change Ben is suggesting here? It'd be really nice to get RESTOREs of non-trivial sizes working again

**Comment by benesch** [comment]
Sure thing!

**Comment by bdarnell** [comment]
Extracted the longer-term fix into #16344
